sdks/uniswapx-sdk/src/order/V3DutchOrder.ts

```diff
@@ -0,0 +1,515 @@
+                            relativeBlocks: output.curve.relativeBlocks,
+                            relativeAmounts: output.curve.relativeAmounts,
+                        },
+                        options.currentBlock
+                    ),
+                }
+            }),
+        };
+    }
+}
+
+function parseSerializedOrder(serialized: string): CosignedV3DutchOrderInfo {
+    const abiCoder = new ethers.utils.AbiCoder();
+    const decoded = abiCoder.decode(V3_DUTCH_ORDER_ABI, serialized);
+    const [
+        [
+            [
+                reactor,
+                swapper,
+                nonce,
+                deadline,
+                additionalValidationContract,
+                additionalValidationData,
+            ],
+            cosigner,
+            [
+                token,
+                startAmount,
+                [inputRelativeBlocks, relativeAmounts],
+                maxAmount,
+            ],
+            outputs,
+            [decayStartBlock, exclusiveFiller, exclusivityOverrideBps, inputOverride, outputOverrides],
+            cosignature,
+        ],
+    ] = decoded;
+
+    return {
+        reactor,
+        swapper,
+        nonce,
+        deadline: deadline.toNumber(),
+        additionalValidationContract,
+        additionalValidationData,
+        cosigner,
+        input: {
+            token,
+            startAmount,
+            curve: {
+                relativeBlocks: decodeRelativeBlocks(inputRelativeBlocks),
+                relativeAmounts,
+            },
+            maxAmount,
+        },
+        outputs: outputs.map(
+            ([token, startAmount, [outputRelativeBlocks, relativeAmounts], recipient]: [
+                string,
+                number,
+                [BigNumber, number[]],
+                string,
+                boolean
+            ]) => ({
+                token,
+                startAmount,
+                curve: {
+                    relativeBlocks: decodeRelativeBlocks(outputRelativeBlocks),
+                    relativeAmounts,
+                },
+                recipient,
+            })
+        ),
+        cosignerData: {
+            decayStartBlock: decayStartBlock.toNumber(),
+            exclusiveFiller,
+            exclusivityOverrideBps: exclusivityOverrideBps,
+            inputOverride: inputOverride,
+            outputOverrides,
+        },
+        cosignature,
+    };
+}
+
+function encodeRelativeBlocks(relativeBlocks: number[]): BigNumber {
+    let packedData = BigNumber.from(0);
+    for (let i = 0; i < relativeBlocks.length; i++) {
+        packedData = packedData.or(BigNumber.from(relativeBlocks[i]).shl(i * 16));
+    }
+    return packedData;
+}
+
+function decodeRelativeBlocks(packedData: BigNumber): number[] {
+    const relativeBlocks: number[] = [];
+    for (let i = 0; i < 16; i++) {
+        const block = packedData.shr(i * 16).toNumber() & 0xFFFF;
+        if (block !== 0) {
```

Comment:
Why is this `0` check here? Idk why we _would_ put a `0` in the relativeBlocks but skipping `0` values seems strange.

Reply:
Maybe instead, we can accept the array length as an argument if this is intended to determine the end.

Reply:
Indeed it was intended to determine the end. If we add an invariant to not allow `0` elements, it'd allow us to easily use `0` to determine the end.

Accepting array length could accomplish this too, but you could also just scan all the bytes to find out how many elements there are without needing outside info passed in.

An edge case that comes to mind with `0` is that it could potentially make sense if it's the first element `relativeBlocks`.
Consider:
`relativeBlocks = [0,2,4]`
`relativeAmounts = [10, 20, 40]`
`startAmount = 100`
This would imply that the order has an **effective** `startAmount` of `110`

But allowing `0` elements to exist also allows for ambiguous cases like `relativeBlocks = [0,0,0,2]` to exist. And allowing it only for the first element feels like a unnecessary edge case to support.

Because of this, I think we should:
Add an invariant to enforce no `0` elements and continue to use it as the delimiter for our encoding
Could also consider an addition in the contract to `revert InvalidDecayCurve()` if we detect this.